volume-group-operator/controllers/volumegroup_controller.go

```diff
@@ -53,14 +55,153 @@ type VolumeGroupReconciler struct {
 //+kubebuilder:rbac:groups="",resources=persistentvolumeclaims/finalizers,verbs=update
 
 func (r *VolumeGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.Log.WithValues("Request.Name", req.Name, "Request.Namespace", req.Namespace)
+
+	instance := &volumegroupv1.VolumeGroup{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+
+			logger.Info("VolumeGroup resource not found")
+
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	vgcObj, err := utils.GetVolumeGroupClass(r.Client, logger, *instance.Spec.VolumeGroupClassName)
+	if err != nil {
+		uErr := utils.UpdateVolumeGroupStatusError(r.Client, instance, logger, err.Error())
+		if uErr != nil {
+			return ctrl.Result{}, uErr
+		}
+		return ctrl.Result{}, err
+	}
+
+	if r.DriverConfig.DriverName != vgcObj.Driver {
+		return ctrl.Result{}, nil
+	}
 
-	// TODO(user): your logic here
+	err = utils.ValidatePrefixedParameters(vgcObj.Parameters)
+	if err != nil {
+		logger.Error(err, "failed to validate parameters of volumegroupClass", "VGClassName", vgcObj.Name)
+		uErr := utils.UpdateVolumeGroupStatusError(r.Client, instance, logger, err.Error())
+		if uErr != nil {
+			return ctrl.Result{}, uErr
+		}
+		return ctrl.Result{}, err
+	}
+	parameters := utils.FilterPrefixedParameters(utils.VolumeGroupAsPrefix, vgcObj.Parameters)
+
+	secretName := vgcObj.Parameters[utils.PrefixedVolumeGroupSecretNameKey]
+	secretNamespace := vgcObj.Parameters[utils.PrefixedVolumeGroupSecretNamespaceKey]
+	secret := make(map[string]string)
+	if secretName != "" && secretNamespace != "" {
+		secret, err = utils.GetSecretData(r.Client, logger, secretName, secretNamespace)
+		if err != nil {
+			uErr := utils.UpdateVolumeGroupStatusError(r.Client, instance, logger, err.Error())
+			if uErr != nil {
+				return ctrl.Result{}, uErr
+			}
+			return reconcile.Result{}, err
```

Comment:
what is the difference between `ctrl.Result{}` to `reconcile.Result{}`, I think we should choose one

Reply:
done